src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -10,11 +10,54 @@
 #define FOH_COMMIT_SIZE (64 * 1024)
 
 FrozenObjectHeapManager::FrozenObjectHeapManager():
-    m_Crst(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC),
+    // This lock is used in both COOP and PREEMP (by profiler) modes
+    m_Crst(CrstFrozenObjectHeap,CRST_UNSAFE_ANYMODE),
+    // This lock is used only in COOP mode
+    m_InitSegmentCrst(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC),
     m_CurrentSegment(nullptr)
 {
 }
 
+void FrozenObjectHeapManager::CreateNewSegment(CrstHolderWithState* activeLock)
+{
+    size_t size = FOH_SEGMENT_DEFAULT_SIZE;
+    if (m_CurrentSegment != nullptr)
+    {
+        // Double the reserved size to reduce the number of frozen segments in apps with lots of frozen objects
+        // Use the same size in case if prevSegmentSize*2 operation overflows.
+        const size_t prevSegmentSize = m_CurrentSegment->GetSize();
+        size = max(prevSegmentSize, prevSegmentSize * 2);
+    }
+
+    const auto segment = new FrozenObjectSegment(size);
+
+    // Release the active lock since segment's initialization logic (RegisterFrozenSegment to be exact) might
+    // hit a gc lock while e.g. profiler is trying to enumerate NonGC objects in GC's callback and waiting for this lock.
+    activeLock->Release();
+
+    bool justInitialized = false;
+    {
+        // We don't want to initialize it twice so a new quick lock is taken.
+        CrstHolder initSegmentLock(&m_InitSegmentCrst);
+
+        // Double-checked locking if some thread already initialized it here.
+        if (!segment->IsInitialized())
+        {
+            justInitialized = true;
+            segment->Initialize();
+        }
+    }
+
+    // Reacquire the active lock
+    activeLock->Acquire();
+    if (justInitialized)
+    {
+        // Update m_CurrentSegment and m_FrozenSegments only for the thread that won the race.
+        m_CurrentSegment = segment;
```

Comment:
This can override a `m_CurrentSegment` that was just allocated by other thread and that has a plenty of space available.

(The implementation that I have sketched in https://github.com/dotnet/runtime/issues/90830#issuecomment-1684841908 tried to avoid this problem by creating the new segment under the main lock, and only taking the other lock for registering the new segment with the GC.)

Reply:
Thanks! I actually didn't realize I can let it allocate objects in a non-registered segment (and register it later before use)